### DISCUSSION TITLE
How graph adjacency impact graph classification?

### DISCUSSION THREAD

**Discussion by evantianzg** [pyg-team/pytorch_geometric#9072] [Q&A]
Hi,

I am performing graph classification using some basic GNN models. I noticed the adjacency of MUTAG for graph classification is just like that:
<img width="1184" alt="截屏2024-03-19 17 08 32" src="https://github.com/pyg-team/pytorch_geometric/assets/22927915/d52dfe2e-7fa7-445b-9657-6db9fda43cfb">

I understand an ideal GNN model for graph-level tasks should be permutation-invariant. So my question is:

Does the order of the adjacency matrix impact the performance of the GNN model? Especially when using the mini-batch strategy. For example, suppose the first graph has "n_1" nodes and "m_1" edges, and the second graph has "n_2" nodes and "m_2" edges. Suppose the batch size is "b". So there should be "b" graphs for each training. In this case, how can we ensure the number of graphs for a mini-batch training is exactly b since all the graph adjacencies are put together in one?

**Comment by rusty1s** [comment]
GNN layers in PyG are usually permutation-invariant, and the order of edges does not matter. `DataLoader` can handle varying sizes across graphs by stacking adjacency matrices diagonally.

The `TUDataset` unwraps the raw format of stacking all graphs into a single file and puts them each into separate `Data` objects.

**Reply by evantianzg** [reply]
@rusty1s Hi, thanks for your replay.

I understand that PyG handles varying sizes across graphs by stacking adjacency matrices in a diagonal fashion. 

I found that there is a "graph_indicator" file in the MUTAG datasets that is used to indicate which graph each node belongs to. I understand the information contained in this file, but I am not sure if this file is used for mini-batch training. Intuitively, I suppose yes. Because if we don't know the correspondences between nodes and graphs, we cannot ensure that graphs can be put into separate `Data` objects.

If I'm right, what should I do with this "graph_indicator" file if a node belongs to multiple graph? Especially when nodes have features.

**Reply by rusty1s** [reply]
Actually, this has nothing to do with data loading/mini-batch training. This file is solely used to separate different graphs in the text file, and we use it to create individual `Data` objects in `TUDataset` for each of this graphs. You can verify this by simply looking at the output of `TUDataset`, e.g.:
```
dataset = TUDataset(..., name='MUTAG')
print(len(dataset))
print(dataset[0])
print(dataset[1])
```
where single graphs are returned.

**Reply by evantianzg** [reply]
Hi @rusty1s Thanks for your reply.

Maybe I didn't express my confusion clearly.

Now I have over 2000 small graphs (with 2 different types of labels), and I want to conduct graph classification using these graphs. I noticed that some nodes belong to multiple graphs. Therefore, when we stack these graph adjacency matrices diagonally, the stacked graph adjacency will not be a diagonal metric: the other part outside the diagonal will not be all 0.

For example, as shown in the image below: on the left is the normal format of the stacked graph adjacency (just like the examples), and on the right is the problem I'm facing now.
<img width="1606" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/22927915/65258fcb-86b5-49e2-9f23-ea6085ecc1b8">

I am not sure if the stacked graph adjacency metric on the right is still fit for graph classification tasks. If so, how can I prepare for it?

**Reply by rusty1s** [reply]
You could use it for graph classification, but you need to be aware that there is a message passing flow now across different graphs, and that your model predictions thus become dependent on which graphs where sampled in your mini-batch.

**Reply by evantianzg** [reply]
Yes, in this case, there will be message passing across different graphs. I understand that independent graphs don't impact model predictions because they are treated as individual samples without common nodes. I intuitively think that the model predictions would be dependent on the graphs sampled in mini-batches when graphs have common nodes. That's why I named the title of this discussion "How graph adjacency impact graph classification?" But I'm still not quite clear about what I should do since there are no such examples and descriptions in PyG documentation.

Do I need to modify the GNN operator that relies on message passing? Or do I just need to duplicate the common nodes and then prepare all the raw files and conduct graph classification tasks as the examples show?

**Reply by rusty1s** [reply]
> "How graph adjacency impact graph classification?"

That's kinda broad :) I am still not sure what you want to do exactly. You can either share edges between graphs in your mini-batch (if you are aware of what you are doing), or you duplicate common nodes in separate graphs so that your graphs stay isolated. In either case, PyG behaves as expected and just performs message passing along the given set of edges. Sorry if I am still having trouble to understand your doubt.

**Reply by evantianzg** [reply]
Sorry for not being clear in my question. You are really nice, and your reply helped me a lot. Here is the full picture.

I am interested in anomaly detection in large-scale graphs. Currently, I am working on the fraud detection of a large-scale transaction network. There are millions of accounts and over ten million transactions between these accounts. Over 1000 accounts are labeled as fraudulent, and the remaining accounts are marked as non-fraudulent. Apparently, it is highly challenging to detect fraudsters via node classification due to the extreme data imbalance. So I tried to build a balanced dataset by combining k-order transaction networks of fraudulent accounts with an equal number of k-order transaction networks from non-fraudulent accounts. Then I could perform graph classification to detect fraudulent accounts. However, I found that these transaction networks had a lot of common neighbors after I carefully checked them.

Intuitively, I think it is ideal not to change these transaction networks (no duplication of common nodes), but that will definitely make the code more complicated (maybe I need to override the `torch_geometric.data.Data.__inc__()` method?). I don't want to get myself into more trouble: I am new to coding :) So I think maybe duplicating the common nodes is the simplest solution.

**Reply by rusty1s** [reply]
I think what you really want to do in this use case is to use `NeighborLoader` and treat this as a node classification problem. Then, you can also pass a custom `sampler` (e.g., `ImbalancedSampler` from PyG) so that fraudulent accounts are sampled more frequently. With this, you can even share nodes across different seed nodes in your mini-batch.

**Reply by evantianzg** [reply]
Never though of such a solution. I am not sure how this method will work in a scenario where the class ratio reaches 1:100, but it sounds great and worth a try.

BTW, I think it is not essential to feed node features into a graph classification model, but would adding node features improve the model's performance?

**Reply by rusty1s** [reply]
If you have meaningful node features, they are always good to add.

**Reply by evantianzg** [reply]
I think so too. There are some distinct differences between fraudulent accounts and non-fraudulent ones. It must be helpful to add node features. Thank you for your suggestions.

Have a good day:)
